lib/Dialect/FIRRTL/Transforms/InferWidths_new.cpp

```diff
@@ -0,0 +1,3510 @@
+
+namespace circt {
+namespace firrtl {
+#define GEN_PASS_DEF_INFERWIDTHS_NEW
+#include "circt/Dialect/FIRRTL/Passes.h.inc"
+} // namespace firrtl
+} // namespace circt
+
+namespace circt {
+
+// 定义Zero结构体
+struct Zero {
+    bool operator==(const Zero&) const { return true; }
+    bool operator==(const FieldRef&) const { return false; }
+        bool operator!=(const Zero&) const { return false; }
+    bool operator!=(const FieldRef&) const { return true; }
+    friend llvm::raw_ostream& operator<<(llvm::raw_ostream& os, const Zero&) {
+        return os << "Zero";
+    }
+};
+
+// 实现FieldRef的operator== !=
+bool FieldRef::operator==(const Zero&) const { 
+    return false; 
+}
+bool FieldRef::operator!=(const Zero&) const { 
+    return true; 
+}
+}
+
+using mlir::InferTypeOpInterface;
+using mlir::WalkOrder;
+
+using namespace circt;
+using namespace firrtl;
+
+//===----------------------------------------------------------------------===//
+// Helpers
+//===----------------------------------------------------------------------===//
+
+static void diagnoseUninferredType(InFlightDiagnostic &diag, Type t,
+                                   Twine str) {
+  auto basetype = type_dyn_cast<FIRRTLBaseType>(t);
+  if (!basetype)
+    return;
+  if (!basetype.hasUninferredWidth())
+    return;
+
+  if (basetype.isGround())
+    diag.attachNote() << "Field: \"" << str << "\"";
+  else if (auto vecType = type_dyn_cast<FVectorType>(basetype))
+    diagnoseUninferredType(diag, vecType.getElementType(), str + "[]");
+  else if (auto bundleType = type_dyn_cast<BundleType>(basetype))
+    for (auto &elem : bundleType.getElements())
+      diagnoseUninferredType(diag, elem.type, str + "." + elem.name.getValue());
+}
+
+/// Calculate the "InferWidths-fieldID" equivalent for the given fieldID + type.
+static uint64_t convertFieldIDToOurVersion(uint64_t fieldID, FIRRTLType type) {
+  uint64_t convertedFieldID = 0;
+
+  auto curFID = fieldID;
+  Type curFType = type;
+  while (curFID != 0) {
+    auto [child, subID] =
+        hw::FieldIdImpl::getSubTypeByFieldID(curFType, curFID);
+    if (isa<FVectorType>(curFType))
+      convertedFieldID++; // Vector fieldID is 1.
+    else
+      convertedFieldID += curFID - subID; // Add consumed portion.
+    curFID = subID;
+    curFType = child;
+  }
+
+  return convertedFieldID;
+}
+
+//===----------------------------------------------------------------------===//
+// Constraint Expressions
+//===----------------------------------------------------------------------===//
+
+/// Check if a type contains any FIRRTL type with uninferred widths.
+static bool hasUninferredWidth(Type type) {
+  return TypeSwitch<Type, bool>(type)
+      .Case<FIRRTLBaseType>([](auto base) { return base.hasUninferredWidth(); })
+      .Case<RefType>(
+          [](auto ref) { return ref.getType().hasUninferredWidth(); })
+      .Default([](auto) { return false; });
+}
+
+using Valuation = std::unordered_map<FieldRef, int32_t, FieldRef::Hash>;
```

Comment:
Can we use llvm::DenseMap instead of unorderd_map? By doing so it wouldn't be necessary to define Zero/Hash structs.